lib/membrane_http_adaptive_stream/sink.ex

```diff
@@ -171,18 +232,34 @@ defmodule Membrane.HTTPAdaptiveStream.Sink do
   end
 
   @impl true
-  def handle_write(Pad.ref(:input, id) = pad, buffer, _ctx, state) do
-    %{storage: storage, manifest: manifest} = state
-    duration = buffer.metadata.duration
+  def handle_write(Pad.ref(:input, track_id) = pad, buffer, ctx, state) do
+    %{
+      target_partial_segment_duration: target_partial_segment_duration,
+      segment_duration: segment_duration
+    } = ctx.pads[pad].options
 
-    {changeset, manifest} =
-      Manifest.add_segment(manifest, id, duration, byte_size(buffer.payload))
+    supports_partial_segments? = target_partial_segment_duration != nil
 
-    state = %{state | manifest: manifest}
+    {changesets, buffers, state} =
+      if supports_partial_segments? do
+        handle_partial_segment(buffer, track_id, segment_duration, state)
+      else
+        handle_segment(buffer, track_id, state)
+      end
 
-    with {:ok, storage} <- Storage.apply_segment_changeset(storage, changeset, buffer.payload),
+    %{storage: storage, manifest: manifest} = state
+
+    with {:ok, storage} <-
+           [changesets, buffers]
+           |> Enum.zip()
+           |> Enum.reduce_while({:ok, storage}, fn {changeset, buffer}, {:ok, storage} ->
+             case Storage.apply_segment_changeset(storage, track_id, changeset, buffer) do
```

Comment:
I think the `Storage` functions should raise instead of returning errors. Not sure if this should be done in this PR, but I think it would simplify things.

Reply:
Sure thing, added an issue for that -> #54